<a href="https://colab.research.google.com/github/solharsh/NLP/blob/master/Introduction_to_BERT.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Top Down Introduction to BERT with HuggingFace and PyTorch

In [5]:
import pickle
import pandas as pd
DATA_PATH = "/content/drive/My Drive/Capstone Project - NLP/Harsh/Project_Checkpoints/"
infile = open(DATA_PATH+'/speech_cleaned_checkpoint.pkl','rb')
df = pickle.load(infile)
import nltk
nltk.download('punkt')
tokenized_corpus = [nltk.word_tokenize(doc) for doc in df.Speech_Cleaned]

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


# HuggingFace and PyTorch

HuggingFace Transformers is an excellent library that makes it easy to apply cutting edge NLP models. I will use their code, such as pipelines, to demonstrate the most popular use cases for BERT. We will need pre-trained model weights, which are also hosted by HuggingFace. I will use PyTorch in some examples.


In [2]:
!pip install transformers -q

     |████████████████████████████████| 645kB 4.8MB/s 
     |████████████████████████████████| 3.8MB 13.5MB/s 
     |████████████████████████████████| 1.1MB 33.5MB/s 
     |████████████████████████████████| 890kB 52.5MB/s 


# What can I use BERT for?

## Text classification

Probably the most popular use case for BERT is text classification. This means that we are dealing with sequences of text and want to classify them into discrete categories.

Here are some examples of text sequences and categories:

    Movie Review - Sentiment: positive, negative
    Product Review - Rating: one to five stars
    Email - Intent: product question, pricing question, complaint, other


In [0]:
from transformers import pipeline, BertTokenizer, BertModel, BertForNextSentencePrediction, BertConfig
import torch

In [4]:
# Text classification - sentiment analysis
nlp = pipeline("sentiment-analysis")

print(nlp("This movie was great!"))
print(nlp("I have just wasted 2 hours of my time."))


[{'label': 'POSITIVE', 'score': 0.9998677372932434}]
[{'label': 'NEGATIVE', 'score': 0.9996421933174133}]


# Named Entity Recognition

In [6]:
# NER / token classification
nlp = pipeline("ner")


{'word': 'Dare', 'score': 0.9987152218818665, 'entity': 'I-PER'}
{'word': '##k', 'score': 0.9988871812820435, 'entity': 'I-PER'}
{'word': 'Warsaw', 'score': 0.9978176355361938, 'entity': 'I-LOC'}


In [7]:
sequence = "My name is Harsh and I live in Bangalore."

for token in nlp(sequence): print(token)

{'word': 'Ha', 'score': 0.9984995126724243, 'entity': 'I-PER'}
{'word': '##rs', 'score': 0.9936491250991821, 'entity': 'I-PER'}
{'word': '##h', 'score': 0.9958865642547607, 'entity': 'I-PER'}
{'word': 'Bangalore', 'score': 0.9970282316207886, 'entity': 'I-LOC'}


# Question Answering

This is the most exciting stuff as per me. Wouldn't it be great if we simply asked a question and got an answer? That is certainly a direction where some of the NLP research is heading (for example T5). 

However, BERT can only handle extractive question answering. It means that we provide it with a context, such as a Wikipedia article, and a question related to the context. BERT will find for us the most likely place in the article that contains an answer to our question, or inform us that an answer is not likely to be found.

In [8]:
# Question Answering
nlp = pipeline("question-answering")

context = "My name is Harsh. I'm Indian. I like to read and watch sci-fi. My home is in Bangalore but I often travel to Delhi. \
My friend, Sagar, lives in Indore."

print(nlp(question="Where does Harsh live?", context=context))
print(nlp(question="Where does Sagar live?", context=context))


{'score': 0.9184085640089847, 'start': 77, 'end': 86, 'answer': 'Bangalore'}
{'score': 0.9569014574200878, 'start': 143, 'end': 149, 'answer': 'Indore.'}


# Other use cases and fine-tuning

There are some other interesting use cases for transformer-based models, such as text summarization, text generation, or translation. BERT is not designed to do these tasks specifically.

# What are the inputs to BERT, and what comes out of it?

Let's start by treating BERT as a black box. The minimum that we need to understand to use the black box is what data to feed into it, and what type of outputs to expect. You can build on top of these outputs, for example by adding one or more linear layers. We can then fine-tune our custom architecture on our data.

# Tokenization
Before you feed your text into BERT, you need to turn it into numbers. That's the role of a tokenizer. Some tokenizers split text on spaces, so that each token corresponds to a word. That would result however in a huge vocabulary, which makes training a model more difficult, so instead BERT relies on sub-word tokenization. Let's see how it works in code.

Each pre-trained model comes with a pre-trained tokenizer (we can't separate them), so we need to download it as well. Let's use it then to tokenize a line of text and see the output.

In [0]:
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')

In [10]:
text = 'I like to read and watch sci-fi.'
tokens = tokenizer.encode(text)
print(tokens)

[101, 1045, 2066, 2000, 3191, 1998, 3422, 16596, 1011, 10882, 1012, 102]


Each token is a number that corresponds to a word (or subword) in the vocabulary. The most frequent words are represented as a whole word, while less frequent words are divided in sub-words. That ensures that we can map the entire corpus to a fixed size vocabulary without unknown tokens (in reality, they may still come up in rare cases). Let's see the length of our model's vocabulary, and how the tokens corresponds to words.


In [11]:
print(f'Length of BERT base vocabulary: {len(tokenizer.vocab)}')
print(f'Text: {text}')
for t in tokens:
  print(f'Token: {t}, subword: {tokenizer.decode([t])}')

Length of BERT base vocabulary: 30522
Text: I like to read and watch sci-fi.
Token: 101, subword: [CLS]
Token: 1045, subword: i
Token: 2066, subword: like
Token: 2000, subword: to
Token: 3191, subword: read
Token: 1998, subword: and
Token: 3422, subword: watch
Token: 16596, subword: sci
Token: 1011, subword: -
Token: 10882, subword: fi
Token: 1012, subword: .
Token: 102, subword: [SEP]


In the example, you can see how the tokenizer split a less common word 'sci-fi' into 2 subwords: 'sci' and 'fi'. BERT tokenizer also added 2 special tokens for us, that are expected by the model: [CLS] which comes at the beginning of every sequence, and [SEP] that comes at the end. [SEP] may optionally also be used to separate two sequences, for example between question and context in a question answering scenario. Another example of a special token is [PAD], we need to use it to pad shorter sequences in a batch, because BERT expects each example in a batch to have the same amount of tokens.


# Outputs

Let's download a pretrained model now, run our text through it, and see what comes out. We will first need to convert the tokens into tensors, and add the batch size dimension (here, we will work with batch size 1).

In [12]:
model = BertModel.from_pretrained('bert-base-uncased')
inputs = torch.tensor(tokens).unsqueeze(0) # Batch size 1
outputs = model(inputs)
print(f'output type: {type(outputs)}, output length: {len(outputs)}')
print(f'first item shape: {outputs[0].shape}')
print(f'second item shape: {outputs[1].shape}')


output type: <class 'tuple'>, output length: 2
first item shape: torch.Size([1, 12, 768])
second item shape: torch.Size([1, 768])


The model outputs a tuple. The first item of the tuple has the following shape: 1 (batch size) x 12 (sequence length) x 768 (the number of hidden units). This is called the sequence output, and it provides the representation of each token in the context of other tokens in the sequence. If we'd like to fine-tune our model for named entity recognition, we will use this output and expect the 768 numbers representing each token in a sequence to inform us if the token corresponds to a named entity.

The second item in the tuple has the shape: 1 (batch size) x 768 (the number of hidden units). It is called the pooled output, and in theory it should represent the entire sequence. It corresponds to the first token in a sequence (the [CLS] token). We can use it in a text classification task - for example when we fine-tune the model for sentiment classification, we'd expect the 768 hidden units of the pooled output to capture the sentiment of the text.

# How was BERT trained?

The models we have been using so far have already been pre-trained, and in some cases fine-tuned as well. What does this actually mean?

Pre-training

In order for a model to solve an NLP task, like sentiment classification, it needs to understand a lot about language. Most of the labelled datasets that we have available are too small to teach our model enough about language. Ideally, we'd like to use all the text we have available, for example all books and the internet. Because it's hard to label so much text, we create 'fake tasks' that will help us achieve our goal without manual labelling.

BERT is trained on a very large corpus using two 'fake tasks': masked language modeling (MLM) and next sentence prediction (NSP). In MLM, we randomly hide some tokens in a sequence, and ask the model to predict which tokens are missing. In NSP, we provide our model with two sentences, and ask it to predict if the second sentence follows the first one in our corpus. The intent of these tasks is for our model to be able to represent the meaning of both individual words, and the entire sentences.

In [13]:
nlp = pipeline("fill-mask")
preds = nlp(f"I am exhausted, it's been a very {nlp.tokenizer.mask_token} day.")
print('I am exhausted, it\'s been a very ***** day.')
for p in preds: print(nlp.tokenizer.decode([p['token']]))
preds = nlp(f"I am excited, it's been a very {nlp.tokenizer.mask_token} day.")
print('I am excited, it\'s been a very ***** day.')
for p in preds: print(nlp.tokenizer.decode([p['token']]))


I am exhausted, it's been a very ***** day.
 busy
 exhausting
 stressful
 taxing
 rough
I am excited, it's been a very ***** day.
 busy
 exciting
 productive
 good
 nice


In [15]:
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
model = BertForNextSentencePrediction.from_pretrained('bert-base-uncased')

first_sentence = "I cut my finger."
second_sentence_right = "The blood started flowing."
second_sentence_wrong = "This website uses cookies."

right = tokenizer.encode_plus(first_sentence, text_pair=second_sentence_right)
wrong = tokenizer.encode_plus(first_sentence, text_pair=second_sentence_wrong)

r1, r2, r3 = torch.tensor(right['input_ids']).unsqueeze(0), torch.tensor(right['token_type_ids']).unsqueeze(0), torch.tensor(right['attention_mask']).unsqueeze(0)
w1, w2, w3 = torch.tensor(wrong['input_ids']).unsqueeze(0), torch.tensor(wrong['token_type_ids']).unsqueeze(0), torch.tensor(wrong['attention_mask']).unsqueeze(0)

right_outputs = model(input_ids=r1, token_type_ids=r2, attention_mask=r3)
right_seq_relationship_scores = right_outputs[0]
wrong_outputs = model(input_ids=w1, token_type_ids=w2, attention_mask=w3)
wrong_seq_relationship_scores = wrong_outputs[0]

print(first_sentence + ' ' + second_sentence_right)
print(f'Next sentence prediction: {right_seq_relationship_scores.detach().numpy().flatten()[0] > 0}')
print(first_sentence + ' ' + second_sentence_wrong)
print(f'Next sentence prediction: {wrong_seq_relationship_scores.detach().numpy().flatten()[0] > 0}')

I cut my finger. The blood started flowing.
Next sentence prediction: True
I cut my finger. This website uses cookies.
Next sentence prediction: False


# Finetuning

As we can see from the examples above, BERT has learned quite a lot about language during pretraining. That knowledge is represented in its outputs - the hidden units corresponding to tokens in a sequence. We can use that knowledge by adding our own, custom layers on top of BERT outputs, and further training (finetuning) it on our own data.


# How does BERT really work?

If training a model is like training a dog, then understanding the internals of BERT is like understanding the anatomy of a dog. It's not required to effectively train a model, but it can be helpful if you want to do some really advanced stuff, or if you want to understand the limits of what is possible.

Let's start by loading up basic BERT configuration and looking what's inside.

In [16]:
config = BertConfig()
config

BertConfig {
  "attention_probs_dropout_prob": 0.1,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 0,
  "type_vocab_size": 2,
  "vocab_size": 30522
}


    768 hidden size is the number of floats in a vector representing each token in the vocabulary
    30522 is the vocabulary size
    We can deal with max 512 tokens in a sequence
    The initial embeddings will go through 12 layers of computation, including the application of 12 attention heads and dense layers with 3072 hidden units, to produce our final output, which will again be a vector with 768 units per token


Let's briefly look at each major building block of the model architecture. We start with the embedding layer, which maps each vocabulary token to a 768-long embedding. We can also see position embeddings, which are trained to represent the ordering of words in a sequence, and token type embeddings, which are used if we want to distinguish between two sequences (for example question and context).

In [17]:
model = BertModel(config)
print(model.embeddings)

BertEmbeddings(
  (word_embeddings): Embedding(30522, 768, padding_idx=0)
  (position_embeddings): Embedding(512, 768)
  (token_type_embeddings): Embedding(2, 768)
  (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
  (dropout): Dropout(p=0.1, inplace=False)
)


Then, we pass the embeddings through 12 layers of computation. This starts with self-attention, is followed by an intermediate dense layer with hidden size 3072, and ends with sequence output that we have already seen above. Usually, we will deal with the last hidden state, i.e. the 12th layer. However, to achieve better results, we may sometimes use the layers below as well to represent our sequences, for example by concatenating the last 4 hidden states.

In [18]:
print(f'There are {len(model.encoder.layer)} layers like this in the model architecture:')
print('---')
print(model.encoder.layer[0])

There are 12 layers like this in the model architecture:
---
BertLayer(
  (attention): BertAttention(
    (self): BertSelfAttention(
      (query): Linear(in_features=768, out_features=768, bias=True)
      (key): Linear(in_features=768, out_features=768, bias=True)
      (value): Linear(in_features=768, out_features=768, bias=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (output): BertSelfOutput(
      (dense): Linear(in_features=768, out_features=768, bias=True)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
  )
  (intermediate): BertIntermediate(
    (dense): Linear(in_features=768, out_features=3072, bias=True)
  )
  (output): BertOutput(
    (dense): Linear(in_features=3072, out_features=768, bias=True)
    (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
    (dropout): Dropout(p=0.1, inplace=False)
  )
)


Finally, we have the pooled output, which is used in pre-training for the NSP task, and corresponds to the [CLS] token hidden state that goes through another linear layer.

In [19]:
print(model.pooler)

BertPooler(
  (dense): Linear(in_features=768, out_features=768, bias=True)
  (activation): Tanh()
)


This is where I learnt it from: 


    Chris Mccormick BERT Research Series on Youtube
    Jay Alammar A Visual Guide to Using BERT for the First Time
    Jay Alammar The Illustrated Transformer
    Peter Bloem Transformers from Scratch
